### PR TITLE
Category links

### PR DESCRIPTION
**Pull request by peter-gerhard**
@schleichardt @lauraluiz please review

### CODE REVIEW

File: product-catalog/app/productcatalog/pages/CategoryLinkDataFactory.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+        return getUrlForLocale(category, locale);
+    }
+
+    private Optional<Locale> getPreferredLocale(LocalizedString localizedString, Collection<Locale> preferredLocales) {
+        return localizedString.getLocales().stream().filter(preferredLocales::contains).findFirst();
+    }
+
+    private Locale getFallbackLocale(LocalizedString localizedString) {
+        return localizedString.getLocales().stream().findFirst().get();
+    }
```
Comment (peter-gerhard) [inline product-catalog/app/productcatalog/pages/CategoryLinkDataFactory.java]: I'll add a test for that

File: conf/templates/home.hbs [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+<head>
+    {{> common/htmlhead}}
+</head>
+
+<body>
+<div class="darkbg hidden"></div>
+{{> common/header}}
+
+<div class="home-page">
+    {{!-- <div class="slick-homepage">
```
Comment (schleichardt) [inline conf/templates/home.hbs]: please remove out commented code

File: product-catalog/app/productcatalog/pages/CategoryLinkDataFactory.java [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+        return category.getName().get(locale);
+    }
+
+    private String getUrl(final Category category) {
+        final Locale locale = getPreferredLocale(category.getSlug(), languages)
+                .orElse(getFallbackLocale(category.getSlug()));
+        return getUrlForLocale(category, locale);
+    }
+
+    private Optional<Locale> getPreferredLocale(LocalizedString localizedString, Collection<Locale> preferredLocales) {
```
Comment (schleichardt) [inline product-catalog/app/productcatalog/pages/CategoryLinkDataFactory.java]: the parameters are not final ;)

File: product-catalog/app/productcatalog/pages/CategoryLinkDataFactory.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+        return preferredLocales.stream().filter(localizedString.getLocales()::contains).findFirst();
+    }
+
+    private Locale getFallbackLocale(LocalizedString localizedString) {
+        return localizedString.getLocales().stream().findFirst().get();
```
Comment (schleichardt) [inline product-catalog/app/productcatalog/pages/CategoryLinkDataFactory.java]: This explodes if no fallback is available.

`localizedString.getLocales().get(0)` is less complex to achieve almost the same behavior.

File: product-catalog/app/productcatalog/pages/CategoryLinkDataFactory.java [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+    }
+
+    private String getUrlForLocale(final Category category, final Locale locale) {
+        return reverseRouter.category(locale.toLanguageTag(), category.getSlug().get(locale), 1).url();
```
Comment (schleichardt) [inline product-catalog/app/productcatalog/pages/CategoryLinkDataFactory.java]: @lauraluiz How should we react if a slug for a locale is not available?

File: product-catalog/test/productcatalog/pages/CategoryLinkDataFactoryTest.java [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+
+                    @Override
+                    public String fragment() {
+                        return null;
+                    }
+                };
+            }
+
+            @Override
+            public Call product(final String locale, final String productSlug, final String sku) {
```
Comment (schleichardt) [inline product-catalog/test/productcatalog/pages/CategoryLinkDataFactoryTest.java]: Why is this necessary?

File: product-catalog/app/productcatalog/pages/CategoryLinkDataFactory.java (hunk 4 above)
Comment (peter-gerhard) [inline product-catalog/app/productcatalog/pages/CategoryLinkDataFactory.java]: can you have a LocalizedString with no entries?
I agree with the second point

File: app/controllers/HomeController.java [hunk 7]
```diff
@@ -32,7 +35,10 @@ public HomeController(final ControllerDependency controllerDependency) {
     }
 
     private Result getResult(final UserContext userContext) {
-        final HomePageContent content = new HomePageContent();
+        final CategoryLinkDataFactory linkFactory = CategoryLinkDataFactory.of(reverseRouter(), userContext.locales());
+        final LinkData menLink = linkFactory.create(categories().findBySlug(Locale.ENGLISH, "men").get());
+        final LinkData womenLink = linkFactory.create(categories().findBySlug(Locale.ENGLISH, "women").get());
+        final HomePageContent content = new HomePageContent(womenLink, menLink);
         return ok(templateService().renderToHtml("home", pageData(userContext, content)));
```
Comment (peter-gerhard) [inline app/controllers/HomeController.java]: There might be a better way to get the categories here, i think for dmexco its ok?!
maybe get the names from config

File: common/app/common/pages/CategoryLinkDataFactory.java [hunk 8]
```diff
@@ -1,23 +0,0 @@
...
-
-    public static CategoryLinkDataFactory of(final List<Locale> locales) {
-        return new CategoryLinkDataFactory(locales);
-    }
-
-    public LinkData create(final Category category) {
-        final String label = category.getName().find(locales).orElse("");
-        return new LinkData(label, "");
-    }
-}
```
Comment (peter-gerhard) [inline common/app/common/pages/CategoryLinkDataFactory.java]: I dropped an already existing CategoryLinkDataFactory...

File: common/test/common/pages/CategoryLinkDataFactoryTest.java [hunk 9]
```diff
@@ -1,24 +0,0 @@
...
-    @Test
-    public void create() {
-        final Category category = categories.findById("5ebe6dc9-ba32-4030-9f3e-eee0137a1274").get();
-
-        final LinkData linkData = CategoryLinkDataFactory.of(singletonList(ENGLISH)).create(category);
-
-        assertThat(linkData.getText()).isEqualTo("TestSnowboard equipment");
-        assertThat(linkData.getUrl()).isEqualTo("");
-    }
-}
```
Comment (peter-gerhard) [inline common/test/common/pages/CategoryLinkDataFactoryTest.java]: and its test

File: product-catalog/app/productcatalog/pages/CategoryLinkDataFactory.java [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+        final Locale locale = getPreferredLocale(category.getName(), languages)
+                .orElse(getFallbackLocale(category.getName()));
+        return category.getName().get(locale);
+    }
+
+    private String getUrl(final Category category) {
+        final Locale locale = getPreferredLocale(category.getSlug(), languages)
+                .orElse(getFallbackLocale(category.getSlug()));
+        return getUrlForLocale(category, locale);
+    }
```
Comment (peter-gerhard) [inline product-catalog/app/productcatalog/pages/CategoryLinkDataFactory.java]: Here is the problem we just talked about. 
1. for the url we need `locale/slug-for-locale` so i need to get the locale first
2. i can not guarantee that the label will have the same locale that the url points to

File: product-catalog/app/productcatalog/pages/CategoryLinkDataFactory.java [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+
+    private Optional<Locale> getPreferredLocale(final LocalizedString localizedString, final Collection<Locale> preferredLocales) {
+        return preferredLocales.stream().filter(localizedString.getLocales()::contains).findFirst();
+    }
+
+    private Locale getFallbackLocale(final LocalizedString localizedString) {
+        return localizedString.getLocales().iterator().next();
+    }
```
Comment (peter-gerhard) [inline product-catalog/app/productcatalog/pages/CategoryLinkDataFactory.java]: it shouldn't be possible that this fails

File: app/controllers/HomeController.java [hunk 12]
```diff
@@ -32,7 +35,10 @@ public HomeController(final ControllerDependency controllerDependency) {
     }
 
     private Result getResult(final UserContext userContext) {
-        final HomePageContent content = new HomePageContent();
+        final CategoryLinkDataFactory linkFactory = CategoryLinkDataFactory.of(reverseRouter(), userContext.locales());
+        final LinkData menLink = linkFactory.create(categories().findBySlug(Locale.ENGLISH, "men").get());
+        final LinkData womenLink = linkFactory.create(categories().findBySlug(Locale.ENGLISH, "women").get());
```
Comment (lauraluiz) [inline app/controllers/HomeController.java]: This is what I meant that eventually (not now) will come from CMS information. For now, I would put in the configuration file (or Messages files, since it will be replaced by CMS) the external ID of the categories, so you fetch them here. This way you don't need the locales.
If it's not found I would send an empty link, but logging an error. But that's not necessary now.

File: product-catalog/app/productcatalog/pages/CategoryLinkDataFactory.java [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+
+    private String getUrl(final Category category) {
+        final Locale locale = getPreferredLocale(category.getSlug(), languages)
+                .orElse(getFallbackLocale(category.getSlug()));
+        return getUrlForLocale(category, locale);
+    }
+
+    private Optional<Locale> getPreferredLocale(final LocalizedString localizedString, final Collection<Locale> preferredLocales) {
+        return preferredLocales.stream().filter(localizedString.getLocales()::contains).findFirst();
```
Comment (lauraluiz) [inline product-catalog/app/productcatalog/pages/CategoryLinkDataFactory.java]: Isn't it the same as
`localizedString.find(preferredLocales)` ?

File: product-catalog/app/productcatalog/pages/CategoryLinkDataFactory.java (hunk 10 above)
Comment (lauraluiz) [inline product-catalog/app/productcatalog/pages/CategoryLinkDataFactory.java]: So my thoughts are: is it that good that we try to recover so badly from a merchant's mistake? The fact that someone is trying to access a category in a language that do not exist shouldn't happen if the project data is well configured. We can make a little mistake a bigger issue if we try to hide it, plus it will create an inconsistent and unpredictable website.
Moreover, I think the responsibility to recover should come from the controller and not from the view model, which should not know how to recover from a particular incident.

File: product-catalog/test/productcatalog/pages/CategoryLinkDataFactoryTest.java [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+import play.mvc.Call;
+
+import java.util.Locale;
+
+import static java.util.Arrays.asList;
+import static java.util.Collections.emptyList;
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class CategoryLinkDataFactoryTest {
+    private final CategoryTree categories = CategoryTree.of(CategoryUtils.getQueryResult("categories.json").getResults());
```
Comment (lauraluiz) [inline product-catalog/test/productcatalog/pages/CategoryLinkDataFactoryTest.java]: It can be `static`.

**Comment by peter-gerhard** [comment]
@lauraluiz ok i refactored the factory assuming that at some point a user will be redirected if she tries to access a language that is not supported and it's a merchants responsibility to keep the data consistent.
can you have a look, pls?

**Comment by lauraluiz** [comment]
Looks good.
As a rule of thumb I'd say we only use all possible locales when it's just displaying some text. That doesn't hurt or alters the navigation.
When it comes to more complex usage of locales, I think it is preferred just one locale to be safe.
